src/adapter/src/rbac.rs

```diff
@@ -51,70 +66,115 @@ impl std::fmt::Display for Attribute {
     }
 }
 
-/// Errors that can occur due to an unauthorized action.
 #[derive(Debug)]
-pub struct UnauthorizedError {
-    /// The action that was forbidden.
-    action: String,
-    kind: UnauthorizedErrorKind,
+pub struct Ownership(ObjectId);
+
+impl Ownership {
+    /// Reports whether a role has ownership over an object.
+    fn check_role(&self, role_id: &RoleId, catalog: &impl SessionCatalog) -> bool {
+        match self.0 {
+            ObjectId::Cluster(id) => catalog.get_cluster(id).owner_id() == *role_id,
+            ObjectId::ClusterReplica((cluster_id, replica_id)) => {
+                catalog
+                    .get_cluster_replica(cluster_id, replica_id)
+                    .owner_id()
+                    == *role_id
+            }
+            ObjectId::Database(id) => catalog.get_database(&id).owner_id() == *role_id,
+            ObjectId::Schema((database_id, schema_id)) => {
+                catalog
+                    .get_schema(
+                        &ResolvedDatabaseSpecifier::Id(database_id),
+                        &SchemaSpecifier::Id(schema_id),
+                    )
+                    .owner_id()
+                    == *role_id
+            }
+            ObjectId::Item(id) => catalog.get_item(&id).owner_id() == *role_id,
+        }
+    }
+
+    /// Reports whether any role has ownership over an object.
+    fn check_roles(&self, role_ids: &BTreeSet<RoleId>, catalog: &impl SessionCatalog) -> bool {
```

Comment:
nit: if possible it would be nice to make `role_ids: impl IntoIterator<Item = &RoleId>` which should allow you to continue to pass in a `&BTreeSet<RoleId>`, as well as any other kind of collection

Reply:
Done